### ISSUE TITLE
pydantic generator does not respect inherited forms

### ISSUE THREAD

**Issue by satra** [linkml/linkml#1432]
**Describe the bug**

The type annotation for relations referring to inherited classes is broken.

**To reproduce**

1. start with this

```yaml
id: https://identifiers.org/brain-bican/kb-model
name: kb-model
prefixes:
  linkml: https://w3id.org/linkml/
  biolink: https://w3id.org/biolink/
  bican: https://identifiers.org/brain-bican/vocab/
  spdx: http://spdx.org/rdf/terms#
  schema: http://schema.org/
  ncbi: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi
imports:
  - linkml:types
  - biolink:biolink-model
default_range: string
default_prefix: bican

classes:
  gene annotation:
    is_a: gene
    slots:
      - referenced in

  genome annotation:
    slots:
      - version

slots:
  referenced in:
    range: genome annotation
```
2. convert to pydantic: `gen-pydantic test.yaml > test.py`
3. the `referenced_in` slot will correctly refer to a `GenomeAnnotation` type
4. now add `is_a: Genome` to genome annotation
5. regenerate, and the `referenced_in` slot becomes a `str`

**Expected behavior**
`referenced_in` slot will continue to refer to a `GenomeAnnotation` type

**Comment by satra** [comment]
btw, the `gen-python` generator seems to do the right thing with the inheritance, so perhaps there is a discrepancy in tests between the two.

**Comment by satra** [comment]
@puja-trivedi and @djarecka - could you please check on our models. this would be great if it's been fixed.

**Comment by puja-trivedi** [comment]
Hi @cmungall, both @djarecka and I tested the code once again with the current version of linkml [1.5.7] as well as the pre released version [v1.5.8rc1] from the main branch and we still run into the same issue. I have attached the exact code we are using below. When we convert to pydantic: `gen-pydantic test.yaml > test.py` the type of `referenced_in` changes from `GenomeAnnotation` to `str`. The key thing to note is that it changes when we add line 23 ( `is_a: genome`) to the **genome annotation** class (line 22). Would you be able to take another look? Thank you.

test.yaml:
`id: https://identifiers.org/brain-bican/kb-model
name: kb-model
prefixes:
  linkml: https://w3id.org/linkml/
  biolink: https://w3id.org/biolink/
  bican: https://identifiers.org/brain-bican/vocab/
  spdx: http://spdx.org/rdf/terms#
  schema: http://schema.org/
  ncbi: https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi
imports:
  - linkml:types
  - biolink:biolink-model
default_range: string
default_prefix: bican

classes:
  gene annotation:
    is_a: gene
    slots:
      - referenced in

  genome annotation:
    is_a: genome
    slots:
      - version

slots:
  referenced in:
    range: genome annotation
`
